core/src/test/java/com/linecorp/armeria/client/HttpClientExpect100HeaderTest.java

```diff
@@ -0,0 +1,492 @@
+import java.util.concurrent.CompletableFuture;
+
+import org.apache.hc.core5.http2.hpack.HPackEncoder;
+import org.apache.hc.core5.util.ByteArrayBuffer;
+import org.jetbrains.annotations.Nullable;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpRequestWriter;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.stream.AbortedStreamException;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.handler.codec.http2.DefaultHttp2HeadersDecoder;
+import io.netty.handler.codec.http2.Http2CodecUtil;
+import io.netty.handler.codec.http2.Http2Flags;
+import io.netty.handler.codec.http2.Http2FrameTypes;
+import io.netty.handler.codec.http2.Http2Headers;
+
+/**
+ * This test is to check the behavior of the HttpClient when the 'Expect: 100-continue' header is set.
+ */
+final class HttpClientExpect100HeaderTest {
+
+    @Nested
+    class AggregatedHttpRequestHandlerTest {
+        @Test
+        void continueToSendHttp1Request() throws Exception {
+            try (ServerSocket ss = new ServerSocket(0)) {
+                final int port = ss.getLocalPort();
+                final WebClient client = WebClient.of("h1c://127.0.0.1:" + port);
+                client.prepare()
+                      .post("/")
+                      .content("foo")
+                      .header(HttpHeaderNames.EXPECT, HttpHeaderValues.CONTINUE)
+                      .execute()
+                      .aggregate();
+
+                try (Socket s = ss.accept()) {
+                    final BufferedReader in = new BufferedReader(
+                            new InputStreamReader(s.getInputStream(), StandardCharsets.US_ASCII));
+                    final OutputStream out = s.getOutputStream();
+                    assertThat(in.readLine()).isEqualTo("POST / HTTP/1.1");
+                    assertThat(in.readLine()).startsWith("host: 127.0.0.1:");
+                    assertThat(in.readLine()).isEqualTo("content-type: text/plain; charset=utf-8");
+                    assertThat(in.readLine()).isEqualTo("expect: 100-continue");
+                    assertThat(in.readLine()).isEqualTo("content-length: 3");
+                    assertThat(in.readLine()).startsWith("user-agent: armeria/");
+                    assertThat(in.readLine()).isEmpty();
+
+                    out.write("HTTP/1.1 100 Continue\r\n\r\n".getBytes(StandardCharsets.US_ASCII));
+
+                    assertThat(in.readLine()).isEqualTo("foo");
```

Comment:
What do you think about adding a check where no data is read between:
```
assertThat(in.readLine()).isEmpty();
```
and
```
out.write("HTTP/1.1 100 Continue\r\n\r\n".getBytes(StandardCharsets.US_ASCII));
```
..?

e.g.

```
s.setSoTimeout(3000); // Raise an exception if no data is read within 3 seconds.
assertThatThrownBy(() -> s.read()).isInstanceOf(SocketTimeoutException.class);
```

.. because otherwise we can't be sure that the content was written before we write the 100 Continue response.

Reply:
For that cases, I use `s.getInputStream().available()` and check if it is 0 or not.
Because if using `read()` and causing a timeout, it could not read the next body properly.